File: day02/dart/day02.dart
import 'dart:io';

void main() {
  List<String> board = ["     ", " 123 ", " 456 ", " 789 ", "     "];
  int rowCoordinate = 1;
  int columnCoordinate = 1;
  // open the file and put each line of the file in a list of strings
  List<String> lines = new File('../input.dat').readAsLinesSync();
  String code = "";

  for (var i = 0; i < lines.length; i++) {
    for (var j = 0; j < lines[i].length; j++) {
      String direction = lines[i][j];
      int dr = 0;
      int dc = 0;
      switch (direction) {
        case 'U':
          dr = -1;
          break;
        case 'D':
          dr = 1;
          break;
        case 'L':
          dc = -1;
          break;
        case 'R':
          dc = 1;
          break;
      }
      if (board[rowCoordinate + dr][columnCoordinate + dc] != ' ') {
        rowCoordinate = rowCoordinate + dr;
        columnCoordinate = columnCoordinate + dc;
      }
    }
    code = code + board[rowCoordinate][columnCoordinate];
  }
  print(code);
}

File: day02/go/day02.go
package main

import (
	"bufio"
	"fmt"
	"os"
	"reflect"
)

func main() {
	board := []string{"     ", " 1234 ", " 456 ", " 789 ", "     "}
	rowCoordinate := 1
	columnCoordinate := 1
	code := ""

	// Open the file.
	theData, _ := os.Open("../input.dat")
	// Create a new Scanner for the file.
	scanner := bufio.NewScanner(theData)
	// Loop over all lines in the file and print them.
	for scanner.Scan() {
		line := scanner.Text()
		fmt.Println(reflect.TypeOf(line).Kind())
		for _, direction := range line {
			// for i := 0; i < len(line); i++ {
			// direction := (string(line[i]))
			dr := 0
			dc := 0

			// direction is in UNICODE value so I have to decode with string()
			switch string(direction) {
			case "U":
				dr = -1
			case "D":
				dr = 1
			case "L":
				dc = -1
			case "R":
				dc = 1
			}
			if string(board[rowCoordinate+dr][columnCoordinate+dc]) != " " {
				rowCoordinate = rowCoordinate + dr
				columnCoordinate = columnCoordinate + dc
			}
		}
		code += string(board[rowCoordinate][columnCoordinate])
	}
	fmt.Println(code)
	fmt.Println(reflect.TypeOf(scanner).Kind())
	fmt.Println(reflect.TypeOf(scanner))

}
